### ISSUE TITLE
segmentation fault on consensus build 

### ISSUE THREAD

**Issue by dominik-handler** [marbl/canu#425]
Hi,

I am still having problems with the consensus build like in this issue:
[https://github.com/marbl/canu/issues/318](url)

When I run it normal,  it tries twice to build it with pbdagcon and I get this error:
```
generatePBDAG()-- retry with bandTolerance = 1200
generatePBDAG()-- Exceeded band size max_k 114 - min_k -2288 = 2402 > band_size = 2400.
generatePBDAG()-- retry with bandTolerance = 2400
generatePBDAG()-- Exceeded band size max_k 1350 - min_k -3452 = 4802 > band_size = 4800.
generatePBDAG()-- Exceeded band size max_k 1191 - min_k -3611 = 4802 > band_size = 4800.
generatePBDAG()-- retry with bandTolerance = 4800
generatePBDAG()-- retry with bandTolerance = 4800
generatePBDAG()-- Exceeded band size max_k 1326 - min_k -3476 = 4802 > band_size = 4800.
generatePBDAG()-- retry with bandTolerance = 4800

Failed with 'Segmentation fault'; backtrace (libbacktrace):
AS_UTL/AS_UTL_stackTrace.C::102 in _Z17AS_UTL_catchCrashiP9siginfo_tPv()
(null)::0 in (null)()
utgcns/libNDFalcon/dw.C::236 in _ZN11NDalignment5alignEPKciS1_iibRNS_13NDalignResultE()
utgcns/libcns/unitigConsensus.C::408 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
./consensus.sh: line 83: 29511 Segmentation fault      $bin/utgcns -G ../ONT-only.${tag}Store/partitionedReads.gkpStore -T ../ONT-only.${tag}Store 1 $jobid -O ./${tag}cns/$jobid.cns.WORKING -maxcoverage 40 -e 0.144 -pbdagcon -threads 8

```

If I try your suggested fix from back then and set it to quick manually, I get this error:
```
Working on unitig 1 of length 35869 (8002 children)
  unitig 1 detected 7853 contains (1265.92x, 94.48%) 149 dovetail (73.93x, 5.52%)
    unitig 1 removing 7853 (1265.92x) contained reads; processing only 0 contained (0.00x) and 149 dovetail (73.93x) reads
utgcns: utgcns/libcns/abAbacus.H:217: abSequence* abAbacus::getSequence(uint32): Assertion `sid < _sequencesLen' failed.

Failed with 'Aborted'; backtrace (libbacktrace):
AS_UTL/AS_UTL_stackTrace.C::102 in _Z17AS_UTL_catchCrashiP9siginfo_tPv()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
utgcns/libcns/abAbacus.H::217 in _ZN8abAbacus11getSequenceEj()
utgcns/libcns/unitigConsensus.C::564 in _ZN15unitigConsensus13generateQuickEP5tgTigPSt3mapIjP6gkReadSt4lessIjESaISt4pairIKjS4_EEEPS2_IjP10gkReadDataS6_SaIS7_IS8_SE_EEE()
utgcns/utgcns.C::505 in main()
(null)::0 in (null)()
../sysdeps/x86_64/start.S::122 in (null)()
(null)::0 in (null)()
./consensus.sh: line 83: 20898 Aborted                 $bin/utgcns -G ../ONT-only.${tag}Store/partitionedReads.gkpStore -T ../ONT-only.${tag}Store 1 $jobid -O ./${tag}cns/$jobid.cns.WORKING -maxcoverage 40 -e 0.144 -quick -threads 8

```

I am using this version:
Canu snapshot v1.4 +161 changes (r8156 6cb65e92d90587caa580df7b1c11c76071447844)

**Comment by skoren** [comment]
There are now two options for consensus, Canu is supposed to automatically try the alternate on the second round. What was the output in your run (it should have run consensus, failed, and on second retry switched) but it seems like it did not?

Try adding ```-edlib``` to the consensus.sh script and see if that fixes it.

**Comment by dominik-handler** [comment]
Yes, it seems the switch is not working.

adding -edlib fixed the whole problem. Out of curiosity, what is this setting doing?

**Comment by skoren** [comment]
The way consensus works is to generate a template of the consensus by pasting together the reads, realigning all the reads to this template to build a DAG, and calling consensus from the DAG. The edlib setting switches the aligner used for the realign step, we've noticed it's more robust than the default but is slower which is why Canu uses it when the default fails.

Can you post your Canu log from unitigging onwards, I'd like to see what it was doing with iterations to see why it didn't switch.

**Comment by dominik-handler** [comment]
Sorry for the question, which log do you exactly mean? Are these the files in the canu-logs folder?

**Comment by skoren** [comment]
Yes, the out files in the canu-logs folder if you're running on the grid or the output to stdout/stderr if you're running on a single system.

**Comment by dominik-handler** [comment]
This is the log for a failed consensus build:
```
canu -p ONT-only -d /project/ngs-work/dhandler/old-run/canu-ONTonly/ genomeSize=150m -nanopore-corrected /project/ngs-work/dhandler/old-run/canu-ONTonly/ONT-only.correctedReads.fasta.gz useGrid=false 
stat(): No such file or directory at /project/ngs-work/dhandler/canu/Linux-amd64/bin/lib/canu/Defaults.pm line 384
stat(): No such file or directory at /project/ngs-work/dhandler/canu/Linux-amd64/bin/lib/canu/Defaults.pm line 384
stat(): No such file or directory at /project/ngs-work/dhandler/canu/Linux-amd64/bin/lib/canu/Defaults.pm line 384
-- Canu snapshot v1.4 +161 changes (r8156 6cb65e92d90587caa580df7b1c11c76071447844)
-- Detected Java(TM) Runtime Environment '1.8.0_92' (from '/software/java/jdk-current/bin/java').
-- Detected gnuplot version '4.6 patchlevel 5' (from 'gnuplot') and image format 'png'.
-- Detected 48 CPUs and 997 gigabytes of memory.
-- Detected Sun Grid Engine in '/software/sge-2011.11/default'.
-- Grid engine disabled per useGrid=false option.
--
-- Run   3 jobs concurrently using   64 GB and  16 CPUs for stage 'meryl'.
-- Run   3 jobs concurrently using   13 GB and  16 CPUs for stage 'mhap (cor)'.
-- Run   6 jobs concurrently using    8 GB and   8 CPUs for stage 'overlapper (obt)'.
-- Run   6 jobs concurrently using    8 GB and   8 CPUs for stage 'overlapper (utg)'.
-- Run  12 jobs concurrently using   20 GB and   4 CPUs for stage 'falcon_sense'.
-- Run  48 jobs concurrently using    4 GB and   1 CPU  for stage 'ovStore bucketizer'.
-- Run  48 jobs concurrently using    8 GB and   1 CPU  for stage 'ovStore sorting'.
-- Run   8 jobs concurrently using    6 GB and   6 CPUs for stage 'read error detection'.
-- Run  48 jobs concurrently using    2 GB and   1 CPU  for stage 'overlap error adjustment'.
-- Run   6 jobs concurrently using   64 GB and   8 CPUs for stage 'bogart'.
-- Run   6 jobs concurrently using   48 GB and   8 CPUs for stage 'consensus'.
--
-- Generating assembly 'ONT-only' in '/project/ngs-work/dhandler/old-run/canu-ONTonly'
--
-- Parameters:
--
--  genomeSize        150000000
--
--  Overlap Generation Limits:
--    corOvlErrorRate 0.3200 ( 32.00%)
--    obtOvlErrorRate 0.1440 ( 14.40%)
--    utgOvlErrorRate 0.1440 ( 14.40%)
--
--  Overlap Processing Limits:
--    corErrorRate    0.5000 ( 50.00%)
--    obtErrorRate    0.1440 ( 14.40%)
--    utgErrorRate    0.1440 ( 14.40%)
--    cnsErrorRate    0.1440 ( 14.40%)
--
--
-- BEGIN ASSEMBLY
--
-- Using fast alignment for consensus (iteration '0').
-- Configured 8 contig and 5 unitig consensus jobs.
-- Finished stage 'consensusConfigure', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting concurrent execution on Tue Mar 28 22:53:42 2017 with 3961.051 GB free disk space (13 processes; 6 concurrently)

    cd unitigging/5-consensus
    ./consensus.sh 1 > ./consensus.000001.out 2>&1
    ./consensus.sh 2 > ./consensus.000002.out 2>&1
    ./consensus.sh 3 > ./consensus.000003.out 2>&1
    ./consensus.sh 4 > ./consensus.000004.out 2>&1
    ./consensus.sh 5 > ./consensus.000005.out 2>&1
    ./consensus.sh 6 > ./consensus.000006.out 2>&1
    ./consensus.sh 7 > ./consensus.000007.out 2>&1
    ./consensus.sh 8 > ./consensus.000008.out 2>&1
    ./consensus.sh 9 > ./consensus.000009.out 2>&1
    ./consensus.sh 10 > ./consensus.000010.out 2>&1
    ./consensus.sh 11 > ./consensus.000011.out 2>&1
    ./consensus.sh 12 > ./consensus.000012.out 2>&1
    ./consensus.sh 13 > ./consensus.000013.out 2>&1

-- Finished on Tue Mar 28 23:19:29 2017 (1547 seconds) with 3961.173 GB free disk space
----------------------------------------
--
-- Running jobs.  Second attempt out of 2.
----------------------------------------
-- Starting concurrent execution on Tue Mar 28 23:19:29 2017 with 3961.173 GB free disk space (4 processes; 6 concurrently)

    cd unitigging/5-consensus
    ./consensus.sh 1 > ./consensus.000001.out 2>&1
    ./consensus.sh 3 > ./consensus.000003.out 2>&1
    ./consensus.sh 4 > ./consensus.000004.out 2>&1
    ./consensus.sh 9 > ./consensus.000009.out 2>&1

-- Finished on Tue Mar 28 23:22:20 2017 (171 seconds) with 3961.173 GB free disk space
----------------------------------------
--
-- 4 consensus jobs failed:
--   job ctgcns/0001.cns FAILED.
--   job ctgcns/0003.cns FAILED.
--   job ctgcns/0004.cns FAILED.
--   job utgcns/0001.cns FAILED.
--
================================================================================
Don't panic, but a mostly harmless error occurred and Canu stopped.

Canu snapshot v1.4 +161 changes (r8156 6cb65e92d90587caa580df7b1c11c76071447844) failed with:
  canu iteration count too high, stopping pipeline (most likely a problem in the grid-based computes)

```

**Comment by brianwalenz** [comment]
This should have been fixed in f4ef049d6018627f6d7be775f55245bf3d10cd15.  Thanks for the log.
